### PR TITLE
feat(subscriptions): Add command to run subscriptions consumer

### PR DESCRIPTION
**Pull request by tkaemming**
This exposes a `snuba subscriptions` command that evaluates subscribed queries as the consumer progresses.

### CODE REVIEW

File: snuba/cli/subscriptions.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    consumer_group: str,
+    auto_offset_reset: str,
+    bootstrap_servers: Sequence[str],
+    max_batch_size: int,
+    max_batch_time_ms: int,
+    schedule_ttl: int,
+    result_topic: Optional[str],
+    log_level: str,
+) -> None:
+    assert result_topic is not None
```
Comment (tkaemming) [inline snuba/cli/subscriptions.py]: I don't think there is a uniformly reasonable default to put here, unless we want subscriptions for all datasets to be routed to the same topic. I guess a reasonable alternative would be to put this on the `Dataset`. I'd like to eventually encapsulate the contents of `snuba.subscriptions` better and expose that as a `dataset.subscriptions` (e.g. `dataset.subscriptions.create(subscription)`, etc.) but that's not strictly required.

File: snuba/cli/subscriptions.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+    setup_logging(log_level)
+    setup_sentry()
+
+    dataset = get_dataset(dataset_name)
+
+    if not bootstrap_servers:
+        bootstrap_servers = settings.DEFAULT_DATASET_BROKERS.get(
+            dataset_name, settings.DEFAULT_BROKERS
+        )
```
Comment (tkaemming) [inline snuba/cli/subscriptions.py]: https://github.com/getsentry/snuba/blob/3378bb1ec69525669669750be0d0d3311a3ae128/snuba/consumers/consumer_builder.py#L51-L56

File: snuba/cli/subscriptions.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    producer = KafkaProducer(
+        {
+            "bootstrap.servers": ",".join(bootstrap_servers),
+            "partitioner": "consistent",
+            "message.max.bytes": 50000000,  # 50MB, default is 1MB
+        },
+        SubscriptionResultCodec(),
+    )
```
Comment (tkaemming) [inline snuba/cli/subscriptions.py]: https://github.com/getsentry/snuba/blob/3378bb1ec69525669669750be0d0d3311a3ae128/snuba/consumers/consumer_builder.py#L88-L94

File: snuba/cli/subscriptions.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        consumer = BatchingConsumer(
+            TickConsumer(
+                SynchronizedConsumer(
+                    KafkaConsumer(
+                        build_kafka_consumer_configuration(
+                            bootstrap_servers,
+                            consumer_group,
+                            auto_offset_reset=auto_offset_reset,
+                        ),
+                        PassthroughCodec(),
```
Comment (tkaemming) [inline snuba/cli/subscriptions.py]: This doesn't need to do any decoding of the message, so this becomes `Consumer[KafkaPayload]`, which meets the `TickConsumer`'s type constraint of `Consumer[Any]`: https://github.com/getsentry/snuba/blob/master/snuba/subscriptions/consumer.py#L76

File: snuba/cli/subscriptions.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+)
+@click.option("--topic")
+@click.option("--partitions", type=int)
+@click.option("--commit-log-topic")
+@click.option(
+    "--commit-log-group",
+    "commit_log_groups",
+    multiple=True,
+    default=["snuba-consumers"],
+)
```
Comment (tkaemming) [inline snuba/cli/subscriptions.py]: https://github.com/getsentry/snuba/blob/3378bb1ec69525669669750be0d0d3311a3ae128/snuba/cli/consumer.py#L24-L28

File: snuba/cli/subscriptions.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+@click.option(
+    "--max-batch-time-ms",
+    default=settings.DEFAULT_MAX_BATCH_TIME_MS,
+    type=int,
+    help="Max length of time to buffer messages in memory before writing to Kafka.",
+)
+@click.option("--schedule-ttl", type=int, default=60 * 5)
+@click.option("--result-topic")
+@click.option("--log-level", default=settings.LOG_LEVEL, help="Logging level to use.")
```
Comment (tkaemming) [inline snuba/cli/subscriptions.py]: Oops, I never updated this to reflect #738, will fix.

File: snuba/cli/subscriptions.py (hunk 1 above)
Comment (tkaemming) [reply snuba/cli/subscriptions.py]: Actually, I guess this should go on the dataset so that this can be part of the bootstrap process.

Comment (tkaemming) [reply snuba/cli/subscriptions.py]: I think this (making this optional by providing the subscription result topic via the dataset) will need to be done later, doing it correctly means moving the `KafkaTopicSpec` object out of the dataset storage module and into streams, and I'd rather take the time to do that correctly instead of rush that out (since this isn't super straightforward to use in development yet anyway): https://github.com/getsentry/snuba/blob/3378bb1ec69525669669750be0d0d3311a3ae128/snuba/datasets/table_storage.py#L16-L20

On the plus side, this probably means taking another look at #746, #747, and #751, all of which felt like hacks anyway.

File: snuba/cli/subscriptions.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    setup_sentry()
+
+    dataset = get_dataset(dataset_name)
+
+    if not bootstrap_servers:
+        bootstrap_servers = settings.DEFAULT_DATASET_BROKERS.get(
+            dataset_name, settings.DEFAULT_BROKERS
+        )
+
+    producer = KafkaProducer(
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/cli/subscriptions.py]: It is inferred from the codec (formatting mine): 

```
snuba/cli/subscriptions.py:121: note: Revealed type is '
snuba.utils.streams.kafka.KafkaProducer[
  Tuple[
    snuba.subscriptions.scheduler.ScheduledTask[
      Tuple[
        snuba.subscriptions.data.SubscriptionIdentifier,
        snuba.subscriptions.data.SubscriptionData,
        fallback=snuba.subscriptions.data.Subscription
      ]  # Subscription
    ],
    typing.MutableMapping[
      builtins.str,
      typing.MutableMapping[builtins.str, Any]
    ],  # ClickhouseQueryResult
    fallback=snuba.subscriptions.worker.SubscriptionResult
  ]  # SubscriptionResult
]'
```

File: snuba/cli/subscriptions.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    commit_log_groups: Sequence[str],
+    consumer_group: str,
+    auto_offset_reset: str,
+    bootstrap_servers: Sequence[str],
+    max_batch_size: int,
+    max_batch_time_ms: int,
+    schedule_ttl: int,
+    result_topic: Optional[str],
+    log_level: Optional[str],
+) -> None:
```
Comment (fpacifici) [inline snuba/cli/subscriptions.py]: It would be useful to add a docstring to this method (which becomes the description of the command when you run it with --help

File: snuba/cli/subscriptions.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+@click.command()
+@click.option(
+    "--dataset",
+    "dataset_name",
+    default="events",
+    type=click.Choice(DATASET_NAMES),
+    help="The dataset to target",
+)
+@click.option("--topic")
+@click.option("--partitions", type=int)
```
Comment (fpacifici) [inline snuba/cli/subscriptions.py]: Do you need to specify the str type for the `topic` option or does click infer it from the type hints of the function? If you do not need to specify that, why did you need to specify that on --partitions ?

File: snuba/cli/subscriptions.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+@click.option(
+    "--consumer-group",
+    default="snuba-subscriptions",
+    help="Consumer group use for consuming the dataset source topic.",
+)
+@click.option(
+    "--auto-offset-reset",
+    default="error",
+    type=click.Choice(["error", "earliest", "latest"]),
+    help="Kafka consumer auto offset reset.",
```
Comment (fpacifici) [inline snuba/cli/subscriptions.py]: Maybe specify that this is for the main consumer and not the commit log consumer ?

File: snuba/cli/subscriptions.py (hunk 3 above)
Comment (fpacifici) [reply snuba/cli/subscriptions.py]: Are you planning to eventually move the logic in common with the consumer_builder somewhere in common, or do you think this should evolve separately ?

File: snuba/cli/subscriptions.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+                            bootstrap_servers,
+                            consumer_group,
+                            auto_offset_reset=auto_offset_reset,
+                        ),
+                        PassthroughCodec(),
+                    ),
+                    KafkaConsumer(
+                        build_kafka_consumer_configuration(
+                            bootstrap_servers,
+                            uuid.uuid1().hex,
```
Comment (fpacifici) [inline snuba/cli/subscriptions.py]: Since this is going to create a new consumer group at every start, do we need to bother garbage collecting them ?

**Review by fpacifici — COMMENTED** [review]
This looks fine, could you pleas e check my comment about removing consumer groups ?

File: snuba/cli/subscriptions.py (hunk 3 above)
Comment (tkaemming) [reply snuba/cli/subscriptions.py]: I am not sure yet. I don't expect this to be the final resting place for this, but I also don't think this belongs in the consumer builder either. (I think it has too many responsibilities at this point.)

My current thinking is to add a `Cluster[TEncoded]` type to `snuba.utils.streams` that exposes `cluster.consumer(…) -> Consumer[TDecoded]` and `cluster.producer(…) -> Producer[TDecoded]`. (The type vars correspond to the [Codec](https://github.com/getsentry/snuba/blob/3378bb1ec69525669669750be0d0d3311a3ae128/snuba/utils/codecs.py#L11-L18) generic.) The configuration will probably be specified by the cluster backend at that point. Some will probably not be overridable (e.g. bootstrap servers come from the cluster definition itself, the partitioner will likely be consistent for all producers to maintain consistency with other implementations, etc.)

This is also related to https://github.com/getsentry/snuba/pull/752#discussion_r375012480.

File: snuba/cli/subscriptions.py (hunk 10 above)
Comment (tkaemming) [reply snuba/cli/subscriptions.py]: I guess so, yeah. I've been mostly considering the commit log consumer an implementation detail that is best left as hidden as possible except when it absolutely needs to be exposed (like the topic itself, since we don't use the default.)

File: snuba/cli/subscriptions.py (hunk 11 above)
Comment (tkaemming) [reply snuba/cli/subscriptions.py]: I did prefix this with c208fc5e0e87e2849fa72bff05bb09650157f9c8 so it's clearer if these show up in metrics where they are coming from.

File: snuba/cli/subscriptions.py (hunk 8 above)
Comment (tkaemming) [reply snuba/cli/subscriptions.py]: Added with 87b6670, tried to keep this short so it wouldn't get truncated by `click`.

File: snuba/cli/subscriptions.py (hunk 10 above)
Comment (tkaemming) [reply snuba/cli/subscriptions.py]: I think I'd prefer to leave this off (it seems too specific with the internal implementation) unless you feel strongly that it's confusing without and/or otherwise necessary.

File: snuba/cli/subscriptions.py [hunk 12]
```diff
@@ -0,0 +1,197 @@
+import signal
+import uuid
+from concurrent.futures import ThreadPoolExecutor
+from contextlib import closing
+from datetime import timedelta
+from typing import Optional, Sequence
+
+import click
```
Comment (fpacifici) [inline snuba/cli/subscriptions.py]: is this intentionally separated from the other module imports (signal, uuid) ?

**Review by fpacifici — APPROVED** [review]

File: snuba/cli/subscriptions.py (hunk 12 above)
Comment (tkaemming) [reply snuba/cli/subscriptions.py]: Yes, this is as dictated by PEP 8:

> Imports should be grouped in the following order:
>
> 1. Standard library imports.
> 2. Related third party imports.
> 3. Local application/library specific imports.

File: snuba/cli/subscriptions.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (wedamija) [reply snuba/cli/subscriptions.py]: <3 mypy

[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/cli/subscriptions.py]: One of us, one of us, one of us
